pkg/virt-launcher/virtwrap/network/podinterface.go

```diff
@@ -40,40 +45,55 @@ var bridgeFakeIP = "169.254.75.1%d/32"
 type BindMechanism interface {
 	discoverPodNetworkInterface() error
 	preparePodNetworkInterfaces() error
+
 	decorateConfig() error
-	loadCachedInterface(name string) (bool, error)
-	setCachedInterface(name string) error
+
+	loadCachedInterface(uid types.UID, name string) (bool, error)
+	setCachedInterface(uid types.UID, name string) error
+
+	// virt-handler that executes phase1 of network configuration needs to
+	// pass details about discovered networking port into phase2 that is
+	// executed by virt-launcher. Virt-launcher cannot discover some of
+	// these details itself because at this point phase1 is complete and
+	// ports are rewired, meaning, routes and IP addresses configured by
+	// CNI plugin may be gone. For this matter, we use a cached VIF file to
+	// pass discovered information between phases.
+	loadCachedVIF(uid types.UID, name string) (bool, error)
+	setCachedVIF(uid types.UID, name string) error
+
+	startDHCP(vmi *v1.VirtualMachineInstance) error
 }
 
 type PodInterface struct{}
 
 func (l *PodInterface) Unplug() {}
 
-func findInterfaceByName(ifaces []api.Interface, name string) (int, error) {
-	for i, iface := range ifaces {
-		if iface.Alias.Name == name {
-			return i, nil
-		}
+func getVifFile(uid types.UID, name string) string {
+	return fmt.Sprintf(vifCacheFile, uid, name)
+}
+
+func writeVifFile(buf []byte, uid types.UID, name string) error {
+	err := ioutil.WriteFile(getVifFile(uid, name), buf, 0644)
+	if err != nil {
+		return fmt.Errorf("error writing vif object: %v", err)
 	}
-	return 0, fmt.Errorf("failed to find interface with alias set to %s", name)
+	return nil
 }
 
-// Plug connect a Pod network device to the virtual machine
-func (l *PodInterface) Plug(vmi *v1.VirtualMachineInstance, iface *v1.Interface, network *v1.Network, domain *api.Domain, podInterfaceName string) error {
-	precond.MustNotBeNil(domain)
+func (l *PodInterface) PlugPhase1(vmi *v1.VirtualMachineInstance, iface *v1.Interface, network *v1.Network, podInterfaceName string) error {
 	initHandler()
 
 	// There is nothing to plug for SR-IOV devices
 	if iface.SRIOV != nil {
 		return nil
 	}
 
-	driver, err := getBinding(vmi, iface, network, domain, podInterfaceName)
+	driver, err := getBinding(vmi, iface, network, nil, podInterfaceName)
```

Comment:
Could you create a new function that would call getBinding with `nil` for Domain? And document this function, how comes we are ok using it without domain -> passing nil to created structures. Or ideally create a new getBinding function that would be just good enough for the usecase in the Phase1. Also, is there a check in driver instances verifying that `domain` is set? Explicitly failing when it is missing would make debugging easier in case we mess something up while moving Phase2 to handler in the future.

Reply:
phase2 already checks domain not nil. Some functions are shared between both phases so we cannot really explicitly check there because domain may or may not be defined in these scenarios.

I've introduced a separate getBinding for phase1 + made it a bit more explicit which end points are for phase2 only.